mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp

```diff
@@ -1100,6 +1101,428 @@ mlir::scf::tileConsumerAndFuseProducersUsingSCF(
+  }
+
+  Location loc = forOp.getLoc();
+  SmallVector<OpFoldResult> offsets = candidateSliceOp.getMixedOffsets();
+  SmallVector<OpFoldResult> sizes = candidateSliceOp.getMixedSizes();
+  SmallVector<OpFoldResult> strides = candidateSliceOp.getMixedStrides();
+  // Check all insert stride is 1.
+  if (llvm::any_of(strides, [](OpFoldResult stride) {
+        return !isConstantIntValue(stride, 1);
+      })) {
+    return rewriter.notifyMatchFailure(
+        candidateSliceOp, "containingOp's result yield with stride");
+  }
+
+  SmallVector<Value> newOuts(forOp.getInits());
+  newOuts.append(dpsInits);
+
+  // Create new scf.for op.
+  rewriter.setInsertionPoint(consumerOp);
+  auto newforOp = rewriter.create<scf::ForOp>(loc, forOp.getLowerBound(),
+                                              forOp.getUpperBound(),
+                                              forOp.getStep(), newOuts);
+  // Move the loop body to the new op.
+  Block *loopBody = forOp.getBody();
+  Block *newLoopBody = newforOp.getBody();
+  rewriter.mergeBlocks(
+      loopBody, newLoopBody,
+      newLoopBody->getArguments().take_front(loopBody->getNumArguments()));
+
+  // Clone the consumer after the insert_slice.
+  rewriter.setInsertionPointAfter(candidateSliceOp);
+  SmallVector<Value> newForOpBlockArgsForConsumerDest;
+  for (unsigned i = loopBody->getNumArguments(),
+                n = newLoopBody->getArguments().size();
+       i < n; i++) {
+    newForOpBlockArgsForConsumerDest.push_back(newLoopBody->getArgument(i));
+  }
+  auto clonedConsumerOp = cast<TilingInterface>(cloneOpAndUpdateDestinationArgs(
+      rewriter, consumerOp, newForOpBlockArgsForConsumerDest));
+
+  // Replace scf.for result's use in the consumer with insert_slice result.
+  rewriter.replaceAllUsesWith(forOp.getResult(0), candidateSliceOp.getResult());
+
+  // Generate the tiled implementation of the consumer of the source.
+  rewriter.setInsertionPoint(candidateSliceOp);
+  FailureOr<TilingResult> tileAndFuseResult =
+      tensor::replaceInsertSliceWithTiledConsumer(
+          rewriter, candidateSliceOp,
+          clonedConsumerOp->getOpOperand(operandNumber));
+  if (failed(tileAndFuseResult)) {
+    return rewriter.notifyMatchFailure(tileableConsumer,
+                                       "failed to tile consumer op: ");
+  }
+
+  // Update the source of the candidateSlice to be the cloned consumer.
+  SmallVector<Value> candidateSliceOpOperands =
+      llvm::to_vector(candidateSliceOp->getOperands());
+  candidateSliceOpOperands[0] = tileAndFuseResult->tiledValues[0];
+  tensor::InsertSliceOp clonedCandidateSliceOp =
+      mlir::clone(rewriter, candidateSliceOp,
+                  candidateSliceOp->getResultTypes(), candidateSliceOpOperands);
+  auto bbArgs = newforOp.getBody()->getArguments();
+  clonedCandidateSliceOp->getOpOperands()[1].set(
+      bbArgs[1 + forOp.getInits().size() + 0]);
+
+  rewriter.replaceAllUsesWith(candidateSliceOp, candidateSliceOp.getSource());
```

Comment:
This cant be right. Did you meant to replace all uses of the `candidateSliceOp` with result of the `clonedCandidateSliceOp`?

Reply:
The state of the IR before this step is :-
```
scf.for {
         . . .
         %matmul = linalg.matmul
         . . .
         %a = tensor.extract_slice %original_insert_slice
         %b = tensor.extract_slice %operand_2_of_consumer
         %c = tensor.extract_slice %operand_3_of_consumer
         . . .
         %tiled_consumer = linalg.generic ins(%a, %b) outs(%c)
         . . .
         %original_insert_slice = tensor.insert_slice %matmul into %iter_arg_for_matmul
         . . .
         %cloned_insert_slice = tensor.insert_slice %tiled_consumer into %iter_arg_for_consumer
         . . .
}
```

To address the `%a` above, I'm replacing uses of `%original_insert_slice` with the source i.e. `%matmul`. This then gives correct use-def chain.

Reply:
Ok, I understand what is happening here. You are cloning the consumer first and then cloning the insert slice. It should be reverse.

1) First clone the insert slice after the original insert slice.
2) Then clone the consumer after it and replace all uses of the loop result with the result of the cloned insert slice. 
At this point you havent changed the program except for creating a clone of the insert_slice -> consumer chain within the loop.
3) Now call the `tensor::replaceInsertSliceWIthTiledConsumer`
4) That returns the tiled result values. You need to use the `getResultTilePosition` method for the consumer operation to figure out what is the shape and offset of the result tile and use that to create the final `tensor.insert_slice` into the destination arg you added to the loop for that result.

This should be the cleanest way of handling this.